python/ray/experimental/serve/queues.py

```diff
@@ -111,304 +148,141 @@ def _serve_metric(self):
                 if backend in self.backend_info:
                     max_batch_size = self.backend_info[backend][
                         "max_batch_size"]
 
-                while len(buffer_queue) and len(work_queue):
-                    # get the work from work intent queue
-                    work = work_queue.popleft()
-                    # see if backend accepts batched queries
-                    if max_batch_size is not None:
-                        pop_size = min(len(buffer_queue), max_batch_size)
-                        request = [
-                            buffer_queue.pop(0) for _ in range(pop_size)
-                        ]
-                    else:
-                        request = buffer_queue.pop(0)
-
-                    work.replica_handle._ray_serve_call.remote(request)
-
-    # selects the backend and puts the service queue query to the buffer
-    # different policies will implement different backend selection policies
-    def _flush_service_queue(self):
-        """
-        Expected Implementation:
-            The implementer is expected to access and manipulate
-            self.queues        : dict[str,Deque]
-            self.buffer_queues : dict[str,sortedlist]
-        For registering the implemented policies register at policy.py!
-        Expected Behavior:
-            the Deque of all services in self.queues linked with
-            atleast one backend must be empty irrespective of whatever
-            backend policy is implemented.
-        """
-        pass
-
-    # _flush function has to flush the service and buffer queues.
-    def _flush(self):
-        self._flush_service_queue()
-        self._flush_buffer()
+                await self._assign_query_to_worker(buffer_queue, worker_queue,
+                                                   max_batch_size)
+
+    async def _assign_query_to_worker(self,
+                                      buffer_queue,
+                                      worker_queue,
+                                      max_batch_size=None):
+
+        while len(buffer_queue) and worker_queue.qsize():
+            worker = await worker_queue.get()
+            if max_batch_size is None:  # No batching
+                request = buffer_queue.pop(0)
+                future = worker._ray_serve_call.remote(request).as_future()
+                asyncio.futures._chain_future(future, request.async_future)
+            else:
+                real_batch_size = min(buffer_queue.qsize(), max_batch_size)
+                requests = [
+                    buffer_queue.pop(0) for _ in range(real_batch_size)
+                ]
+                future = worker._ray_serve_call.remote(requests).as_future()
+                future.add_done_callback(
+                    _make_future_unwrapper(
+                        client_futures=[req.async_future for req in requests],
+                        host_future=future))
 
 
 class CentralizedQueuesActor(CentralizedQueues):
```

Comment:
Idon't think we need this wrapper class anymore?

Reply:
@simon-mo can we directly use `CentralizedQueues` instead of `CentralizedQueuesActor` ?